password/src/Data/Password/Validate.hs

```diff
@@ -126,122 +143,139 @@ data CharacterCategory
   -- ^ ASCII digits
   deriving (Eq, Ord, Show)
 
--- | Convert 'CharacterCategory' into associated predicate function
+-- | Convert a 'CharacterCategory' into its associated predicate function
+--
+-- @since 2.1.0.0
 categoryToPredicate :: CharacterCategory -> (Char -> Bool)
 categoryToPredicate = \case
   Uppercase -> isAsciiUpper
   Lowercase -> isAsciiLower
   Special -> isSpecial
   Digit -> isDigit
 
--- | Possible reason of 'Password' being invalid
+-- | Possible reason for a 'Password' to be invalid.
+--
+-- @since 2.1.0.0
 data InvalidReason
   = PasswordTooShort !Int !Int
   -- ^ Length of 'Password' is too short.
   --
-  -- Expected at least 'Int' characters but the actual length is 'Int'
+  -- @PasswordTooShort expected found@
```

Comment:
It is not immediately clear what `expected` and `found` are here, although I don't really have a good idea of a way to make this documentation, or the datatype more clear.

An easy thing to do would be to add type aliases:

```haskell
-- | The expected (minimum or maximum) number of characters.
type ExpectedLen = Int

-- | The found (incorrect) number of characters
type FoundLen = Int

data InvalidReason = PasswordTooShort ExpectedLen FoundLen
```

Reply:
How about now?

Reply:
Yeah, I think this looks good now!